core_lib/tool/penciltool.cpp

```diff
@@ -229,14 +236,14 @@ void PencilTool::paintAt( QPointF point )
     Layer* layer = mEditor->layers()->currentLayer();
     if ( layer->type() == Layer::BITMAP )
     {
-        qreal opacity = 1.0f;
+        qreal opacity = 1.0;
         mCurrentWidth = properties.width;
         qreal brushWidth = mCurrentWidth;
 
         BlitRect rect;
 
         rect.extend( point.toPoint() );
-        mScribbleArea->drawPencil(point,
+        mScribbleArea->drawPencil( QPoint( qRound(point.x() ), qRound(point.y() )),
```

Comment:
@CandyFace  Did you remember why you replace the QPointF with QPoint?

Reply:
Thinking in context of pixels, you can't have a half pixel, so there should be no reason why the point you're drawing is floats.

I believe that's why I changed it.

Reply:
Just did some small tests. Please check the results as below:
Zoom 845.6%, Pencil width 2
![](https://imgur.com/XcveJj1.png)

The left image is using QPointF, the Right image is using QPoint.

In the right image, the stroke has some "sudden moves", which is the timing of floating numbers just cross the 0.5, for example, from 1.49 to 1.52. It's a very small move, but causes a whole pixle shift after rounding to integers.

Since Qt it quite ok to deal with floating-point positions. I think we should go back to QPointF. I'll do it later, just let you know. :)

Reply:
Ah yes my mistake, it's affecting the antialiasing of the brush... I didn't think about that when I made the change. Do change it back to floating-points ;)

Reply:
Just a reference: 2226afd16406f0d9d50df6390e29c22e89392dd9